docs/architecture/tree-chunking.md

```diff
@@ -0,0 +1,123 @@
+
+### MisIndexed Chunks
+
+Suppose an honest chunker follows a strategy creating chunks 1,2,3,4,5.
+Suppose an honest chunkee requests chunk 3 from a malicious chunker,
+who responds with chunk 4. The chunkee applies chunk 4 to its Merkle tree
+and verifies that it's correct. At this point the chunkee believes it
+successfully applied chunk 3. This raises two questions:
+
+- When the chunkee actually requests and receives chunk 4, and then discovers it already
+  applied that chunk, how will it know which peer was malicious, the original
+  one or the new one?
+- How will the chunkee discover that it is missing the real chunk 3?
+
+### NonIndexed Chunks
+
+Suppose an honest chunker follows a strategy creating chunks 1,2,3,4,5.
+Suppose a malicious chunker follows an alternative strategy, creating chunks
+Q,R,S,T,U, none of which correspond to chunks 1,2,3,4,5, but all of which are
+valid chunks in the tree.
+Suppose an honest chunkee requests chunk 3 from the malicious chunker,
+who responds with chunk R. The chunkee applies chunk R to its Merkle tree and
+verifies that it's correct. At this point the chunkee believes it
+successfully applied chunk 3. This raises two questions:
+
+- When the chunkee receives another chunk that overlaps with chunk R, how will
+  it know which peer was malicious, the original one or the new one?
+- How will the chunkee determine which real chunks it is missing?
+
+## Provably Indexed Chunking
+
+An alternative approach would be a strategy that chunks the tree in a manner
+such that all chunks have a verifiable index. Certain tree structures may be
+able to incorporate this requirement into their design. In the extreme case,
+the map from chunks to indices (or vice versa) could be committed into the tree
+itself. But generally, we would like to avoid such requirements on the tree and
+state and instead pursue a general strategy for chunking trees that preserves
+verifiable chunk indices.
+
+Consider an example strategy for illustration of how this would work:
+
+- the first chunk (chunk 0) contains all nodes in the top 10 layers of the tree
+- the nodes in the layer 10, which are part of chunk 0, form the roots for 1024
+  sub-trees that contain the rest of the tree
+- each of these sub-trees is its own chunk, numbered from 1 to 1024
+- so long as chunk 0 is received first, each additional chunk can be verified
+  with its index. For instance, if we receive chunk 5, the root hash of the
+  sub-tree contained therein should correspond to the 5th node in layer 10
+
+While this example provides the intuition for how the design might work,
+we now need to generalize it to handle arbitrary tree sizes and bounded
+chunk sizes. With the example above, we were only required to receive chunk 0 up
+front, and could then receive the other chunks in any order. However to
+generalize, we may be required to receive multiple chunks in order first, before
+we get to a point where we can receive them in any order.
+
```

Comment:
For a depth-first traversal, consider the following:

- Chunkee & chunker negotiate a manifest `m` mapping chunk indices to contiguous subsets (ranges) of keyspace. They agree on `n` chunks, and they now both can map the integer range `0 - n - 1` to a subset of keyspace (start key, end key), where the end keys and start keys of contiguous chunks are identical (end is exclusive).
- Chunkee requests chunk `c_k` from chunker. Chunker sends some `c_l` where `k` might or might not be equal to `l`.
- Chunkee checks keys in `c_l` against known mapping from `k`. If keys are not within the range, chunkee rejects chunk & bans peer. Chunkee now knows that the keys do belong to chunk `c_k`, although it does not know if the values are correct.
- Chunkee now chooses random key in the chunk and requests Merkle proof of key, value pair (against known state root from light client) from chunker, or chooses random key range not in the included keys but in the chunk range and requests range proof of non-inclusion from chunker. If Merkle proof is not provided, chunkee rejects chunk and bans peer. If Merkle proof does not validate, chunkee rejects chunk and bans peer.
- Repeat random requests some `r` times with `r` a security parameter chosen according to expected peer behaviour, total number of chunks, etc.
- Chunkee now knows (in expectation) that this chunk is valid - that it has the correct keys and all the keys within this range. Chunkee applies chunk.
- Repeat for all chunks in range `0 - n - 1`.

- At the end, verify whole tree. If the state root does not match, binary search by requesting Merkle proofs of subtrees from peers and fetch any incorrect chunks. This step could be taken earlier if we're concerned about malicious peers.

For trees which can handle out-of-order-insertion, that should work in parallel as far as I can tell.

Reply:
It looks like this still doesn't solve all connected chunker are the malicious (but provide consistent chunks) case. 

If at first, the chunkee negotiates with 3 malicious peers, then it has no chance to connect to correct network.
If at first, the 2 peers send chunkee a set of chunk and keyspace mapping while other 2 peers send chunkee another kind of mapping, how the chunkee would behave (decide which 2 peers are malicious)?

If this is not the case merkle proof to resolve, I think negociate hash (which provides more information than indexes) of chunk (begin and end key can still be kept) would be easier. 
I cannot see the additional benefit merkle proof provides than hash of each chunk. Hash verification is:
1. much more efficient (without proof request-responde round trip)
2. rejecting unexpected chunks earlier.
3. doesn't need top layer chunks comes in order to achieve a verifiable tree, all chunks can be processed as long as it arrives

Reply:
> * Chunkee now chooses random key in the chunk and requests Merkle proof of key, value pair (against known state root from light client) from chunker

why choose random key for existence check? Wouldn't one whole range request be better (avoid `c_l` miss or add key but still in range)

Reply:
> If at first, the 2 peers send chunkee a set of chunk and keyspace mapping while other 2 peers send chunkee another kind of mapping, how the chunkee would behave (decide which 2 peers are malicious)?

Different keyspace mappings could be valid, there doesn't need to be a single canonical one - if the chunkee and chunker can't agree on the manifest (maybe the chunkee wants smaller or larger chunks), the chunkee can disconnect.

> If this is not the case merkle proof to resolve, I think negociate hash (which provides more information than indexes) of chunk (begin and end key can still be kept) would be easier.
I cannot see the additional benefit merkle proof provides than hash of each chunk.

The Merkle proof provides existence / non-existence proofs against a trusted state root - how would the chunkee know that a chunk with a particular hash was in fact the correct chunk for the keyspace?

> why choose random key for existence check? Wouldn't one whole range request be better (avoid c_l miss or add key but still in range)

Just so the proof construction, bandwidth, and verifications costs are lower. You could request proofs for the whole chunk, it would be more expensive.

Reply:
> how would the chunkee know that a chunk with a particular hash was in fact the correct chunk for the keyspace?

later when the chunkee compares root hash with one in the block's header?

Reply:
> later when the chunkee compares root hash with one in the block's header?

Sure, but then we can't validate chunks as we receive them, so we might waste time receiving & storing a bunch of chunks which later turn out to be invalid.

Reply:
How hard it is to construct the chunk with **a particular hash** with invalid data inside?

Reply:
> Different keyspace mappings could be valid, there doesn't need to be a single canonical one - if the chunkee and chunker can't agree on the manifest (maybe the chunkee wants smaller or larger chunks), the chunkee can disconnect.

Then how chunkee knows keyspace at first?

Reply:
> Then how chunkee knows keyspace at first? Without fixed-size chunk, you cannot support eager state sync right?

Chunker and chunkee negotiate on a manifest mapping integer chunk indices to bounded subsets of (the entire) keyspace. What do fixed-size chunks have to do with eager state sync? The chunker and chunkee would also need to pick a block height of the state to sync, if that's what you mean.

Reply:
> Chunker and chunkee negotiate on a manifest mapping integer chunk indices to bounded subsets of (the entire) keyspace.

I mean chunkee is brand new node, it doesn't have any information about how many keys in total. `For entire keyspace` did you mean whole possible 32 bytes of sha256 (hash of iavl tree node) i.e. 0x10..00 - 0xFF..FF?



> What do fixed-size chunks have to do with eager state sync? The chunker and chunkee would also need to pick a block height of the state to sync, if that's what you mean.

I mean think about there are 100 chunkee peers and 1 chunker. All 100 chunkee want negociate different chunk size of chunker. i.e. The first chunkee wants chunk size be 1M, the second one wants 2M, the third one wants 3M, ..., the 100th chunkee want 100M. To serve 100 kinds of clients, if the chunker want eagerly prepared the chunks (without traverse ival tree each time a request comes), does it need prepare 100 kinds of chunks on disk?

Reply:
> I mean chunkee is brand new node, it doesn't have any information about how many keys in total. For entire keyspace did you mean whole possible 32 bytes of sha256 (hash of iavl tree node) i.e. 0x10..00 - 0xFF..FF?

Yes; the manifest maps integer indices to contiguous subsets of the entire keyspace.

> I mean think about there are 100 chunkee peers and 1 chunker. All 100 chunkee want negotiate different chunk sizes of chunker. i.e. The first chunkee wants chunk size to be 1M, the second one wants 2M, the third one wants 3M, ..., the 100th chunkee want 100M. To serve 100 kinds of clients, if the chunker wants eagerly prepared the chunks (without traverse ival tree each time a request comes), does it need to prepare 100 kinds of chunks on disk?

That's right, but there could be suggested chunk sizes (which could themselves change over time); it's all in the peer-to-peer protocol and altruism is required anyways.